locator/abstract_replication_strategy.cc

```diff
@@ -165,19 +165,35 @@ effective_replication_map::get_ranges(inet_address ep) const {
     });
 }
 
-// Caller must ensure that token_metadata will not change throughout the call if can_yield::yes.
+// Caller must ensure that token_metadata will not change throughout the call.
 future<dht::token_range_vector>
 abstract_replication_strategy::get_ranges(inet_address ep, token_metadata_ptr tmptr) const {
+    co_return co_await get_ranges(ep, *tmptr);
+}
+
+// Caller must ensure that token_metadata will not change throughout the call.
+future<dht::token_range_vector>
+abstract_replication_strategy::get_ranges(inet_address ep, const token_metadata& tm) const {
     dht::token_range_vector ret;
-    const auto& tm = *tmptr;
+    if (!tm.is_normal_token_owner(ep)) {
+        co_return ret;
+    }
     const auto& sorted_tokens = tm.sorted_tokens();
     if (sorted_tokens.empty()) {
         on_internal_error(rslogger, "Token metadata is empty");
     }
     auto prev_tok = sorted_tokens.back();
     for (auto tok : sorted_tokens) {
-        auto eps = co_await calculate_natural_endpoints(tok, tm);
-        if (eps.contains(ep)) {
+        bool should_add = false;
+        if (get_type() == replication_strategy_type::everywhere_topology) {
+            // everywhere_topology deserves this special fast path.
+            // Using the common path would make the function quadratic in the number of endpoints.
+            should_add = true;
+        } else {
+            auto eps = co_await calculate_natural_endpoints(tok, tm);
+            should_add = eps.contains(ep);
+        }
+        if (should_add) {
             insert_token_range_to_sorted_container_while_unwrapping(prev_tok, tok, ret);
```

Comment:
Out of scope for this series, but insert_token_range_to_sorted_container_while_unwrapping is overly complicated for this caller.  Since we're traversing the sorted_tokens I'm pretty sure we can handle the end-cases before/after the loop and use much simpler logic to insert the sorted token ranges into the returned vector.